src/libstd/sys/unix/mutex.rs

```diff
@@ -67,3 +68,38 @@ impl Mutex {
         debug_assert!(r == 0 || r == libc::EINVAL);
     }
 }
+
+pub struct ReentrantMutex { inner: UnsafeCell<ffi::pthread_mutex_t> }
+
+unsafe impl Send for ReentrantMutex {}
+unsafe impl Sync for ReentrantMutex {}
+
+impl ReentrantMutex {
+    pub unsafe fn new() -> ReentrantMutex {
+        let mut mutex: ffi::pthread_mutex_t = mem::uninitialized();
```

Comment:
Unfortunately this may also not be valid on all platforms, I believe once a mutex has been initialized with `pthread_mutex_init` it's not allowed to change addresses. I think that iOS doesn't play well if this happens for example. Perhaps this could have a separate function to initialize it after it's been boxed by a call above?